### ISSUE TITLE
Add Constrained Delaunay Triangulation

### ISSUE THREAD

**Issue by wez** [Stoeoef/spade#12]
Any plans on adding CDT support to this library?

**Comment by Stoeoef** [comment]
Hey,

thanks for the input. I've looked through the usual algorithm for constructing the CDT, it doesn't seem to be too complicated since many parts of the algorithm are already implemented. So, I'll try to add it, although I cannot give any estimates about when it's going to be done yet - I'll keep this issue updated.

**Comment by wez** [comment]
Thanks!

**Comment by Stoeoef** [comment]
I've just pushed the cdt branch, see 148259eceb33f70053a99bb51a0ecb185a7aefbe . It still has some rough edges, the public interface might also slightly change in the next weeks, but the CDT seems to work so far.
Notable restrictions:
 - It only supports constraint edges that do not cross each other. They may intersect at their end points. It will panic if intersecting edges are added.
 - Adding constraints to a degenerate triangulation will panic - add at least three vertices before adding constraints.

If you want to try it out, add this to your dependencies:
`spade = { git = "https://github.com/Stoeoef/spade.git", branch = "cdt" }`

Let me know if some feature is still missing, I'll try to add it in the upcoming 1.2 release.

**Comment by wez** [comment]
Thanks!  I'm not likely to be have time to try this out for a week or two, but am thankful that you added this!

**Comment by Stoeoef** [comment]
I'll be adding this in the next version update... things to do until then, just so I won't forget anything:

- [ ] More test cases, especially some tests that add dependency edges and normal vertices interleaved
- [ ] Update the handling of degenerate triangles: Their points should be connected by edges, operations like vertex location should be supported, although possibly with O(n) performance. This change will it make possible to generate a cdt by only inserting constraint edges.
- [ ] Test cases for the previous feature
- [ ] Probably create an example covering cdt's.
- [ ] Get rid of some code duplication: I've copied the parts handling degenerate triangulations from the 2D Delaunay triangulation, this must be put together
- [ ] (Optional) At least consider supporting vertex removal in cdt's, I'm not sure how difficult this will be. The algorithm needs to keep constraint edges consistent when removing edges from the graph.
- [ ] Documentation of all new features
- [ ] (Optional) Add a method that checks if a given edge intersects a constraint edge.

**Comment by wez** [comment]
<img width="812" alt="screen shot 2017-09-03 at 2 10 19 pm" src="https://user-images.githubusercontent.com/117777/30006637-a9e09bc2-90b1-11e7-91a7-fbd80f125aee.png">

I finally got around to integrating and trying out this code.  I'm still a bit of a n00b when it comes to spatial and graph stuff, so I'm not sure if what I'm seeing in the attached screen shot is a bug or if I'm doing something wrong.

My use case is applying the CDT to a PCB track routing problem.  In the attached screenshot the brown lines are the edges extracted from the CDT.  The red features are obstacles that should not be crossed; the exterior of those shapes is added to the CDT as constrained edges.  If you zoom in on those red circles you can see that there are some brown lines crossing through the interior of the red circles.

My expectation was that the constrained edges would walk around the circumference of the circle rather than sometimes bridging across it.  Not every vertex from the circle crosses in this way, which makes me wonder if the triangulation is contributing to this.  Is my expectation just dead wrong?

Aside from this it works well provided that unconstrained points have been added before adding any constraints, otherwise this line panics:

https://github.com/Stoeoef/spade/blob/cdt/src/delaunay/cdt.rs#L381

I think that's reasonable to cover in the docs once this feature graduates to a release.

**Comment by urschrei** [comment]
@wez I'm curious as to how you managed to get the constrained edges out. I can build the CDT, but I'm a bit stuck as to how to do anything with it, beyond using `is_constraint_edge`.

**Comment by wez** [comment]
@urschrei I'm walking all edges and populating a petgraph graph so that I can use its graph algorithms on the result:

```
use petgraph::graphmap::{UnGraphMap, NodeTrait};
use spade::delaunay::Subdivision;

pub type CDT<Value> = ConstrainedDelaunayTriangulation<Vertex<Value>, FloatKernel>;
pub type CDTUnGraph<Value> = UnGraphMap<Vertex<Value>, f64>;

pub fn cdt_to_graph<Value, F>(cdt: &CDT<Value>, mut edge_cost: F) -> CDTUnGraph<Value>
    where Value: NodeTrait,
          F: FnMut(&EdgeHandle<Vertex<Value>>) -> Option<f64>
{
    let mut g = CDTUnGraph::<Value>::new();

    for edge in cdt.edges() {
        if let Some(cost) = edge_cost(&edge) {
            let from = &*edge.from();
            let to = &*edge.to();

            g.add_node(*from);
            g.add_node(*to);
            g.add_edge(*from, *to, cost);
        }
    }

    g
}
```

The resultant graph doesn't need to know whether an edge is constrained or not, just whether it is present.  I care about the extraneous edges that I mentioned above in my application, so the `edge_cost` function I pass to `cdt_to_graph` has some logic to test each edge for collision against the obstacles and elide the "bad" edges by returning `None`.

**Comment by wez** [comment]
Also worth mentioning: in my application, I'm buffering/inflating the polygon that outlines the obstacle so that I have a valid path just outside it.  That way the nodes and edges all represent valid candidate destinations and paths.  That is why I don't need to know whether a given edge is constrained or not.

**Comment by urschrei** [comment]
Ah, gotcha. I need to get the infinite face of the CDT, so I can iterate and tag each triangle as "interior" or "exterior" (I'm trying to determine minimum distance between non-convex polygons, and I think "exterior" triangles will have at least one infinite edge, and exactly two vertices belonging to the outer polygon ring). Then I can test the outer triangles for containment against all the vertices in the other polygon, keeping the minimum distance from a contained vertex to the triangle edges and vertices if I get a hit. Flipping the polygons and repeating the whole thing should give me the minimum distance.

**Comment by Stoeoef** [comment]
@wez I finally found some time again to work on spade, sorry for the delay.
To answer your question about the cdt-lines crossing your obstacles: I think the triangulation is actually doing what it is expected to do. A CDT is still a triangulation, that means, every area within the convex hull must be covered by a triangle which is the case for the area within the red circles.
However, there might be some workarounds for your specific problem. It would not be too difficult to extract the subgraph that contains all edges except those inside the circles. Here's an algorithm that would do that by "filling out the interior of the board":
- Take an arbitrary face adjacent to the infinite face and add it to a todo list
- While the todo list is not empty: Pop a face, add it to a result list, and add its three neighboring faces to the todo list only if two conditions hold:
  - the face is not contained in the result list
  - the edge that is shared by these faces is no constraint edge

The result list will now contain all faces in the interior of your board.

**Comment by Stoeoef** [comment]
Implemented with 180286ce6205ed29d7083eb394ad142568918f09

**Comment by wez** [comment]
Thanks!
